File: exemplos/python/estruturas_dados/estruturas_dados_simples.py
"""
ESTRUTURAS DE DADOS SIMPLES - Python
Homogêneas, Heterogêneas e Referências
Cada operação comentada com sua complexidade
"""

import time
from typing import List, Dict

# ====================================
# 1. ESTRUTURAS HOMOGÊNEAS (LISTAS)
# ====================================

def exemplo_lista_basica():
    """Exemplo com lista homogênea em Python"""
    print("\n=== LISTA HOMOGÊNEA ===")
    
    # Criação de lista - O(n)
    numeros = [10, 20, 30, 40, 50]                      # O(n) - cria lista com n elementos
    print(f"Lista criada: {numeros}")                   # O(n) - impressão da lista
    
    # Acesso direto - O(1)
    valor = numeros[2]                                   # O(1) - acesso por índice
    print(f"Elemento na posição 2: {valor}")            # O(1) - impressão
    
    # Busca linear - O(n)
    procurado = 30
    posicao = -1
    for i, num in enumerate(numeros):                    # O(n) - percorre lista
        if num == procurado:                             # O(1) - comparação
            posicao = i                                  # O(1) - atribuição
            break                                        # O(1) - saída do loop
    
    if posicao != -1:                                    # O(1) - verificação
        print(f"Valor {procurado} encontrado na posição {posicao}")
    else:
        print(f"Valor {procurado} não encontrado")
    
    # Busca com método built-in - O(n)
    try:
        pos_builtin = numeros.index(30)                  # O(n) - busca interna
        print(f"Usando .index(): posição {pos_builtin}")
    except ValueError:
        print("Valor não encontrado")

def operacoes_lista_simples():
    """Operações básicas com listas"""
    print("\n=== OPERAÇÕES COM LISTA ===")
    
    numeros = []                                         # O(1) - lista vazia
    
    # Adição no final - O(1) amortizado
    print("Adicionando elementos:")
    for i in range(5):                                   # O(n) - loop 5 vezes
        numeros.append(i * 10)                           # O(1) - adiciona no final
        print(f"Adicionado {i * 10}, lista: {numeros}") # O(n) - impressão da lista
    
    # Soma dos elementos - O(n)
    soma = sum(numeros)                                  # O(n) - percorre toda lista
    print(f"Soma dos elementos: {soma}")
    
    # Remoção - O(n) no pior caso
    if numeros:                                          # O(1) - verificação se não vazia
        removido = numeros.pop(2)                        # O(n) - remove índice 2, move elementos
        print(f"Removido elemento {removido}, lista: {numeros}")

# ====================================
# 2. ESTRUTURAS HETEROGÊNEAS (CLASSES/DICTS)
# ====================================

class Funcionario:
    """Classe simples para representar funcionário"""
    
    def __init__(self, nome: str, idade: int, salario: float):
        """Construtor - O(1) para atribuições numéricas, O(n) para strings"""
        self.nome = nome                                 # O(1) - atribuição de referência
        self.idade = idade                               # O(1) - atribuição direta  
        self.salario = salario                           # O(1) - atribuição direta
    
    def __str__(self):
        """Representação string - O(1)"""
        return f"{self.nome}, {self.idade} anos, R$ {self.salario:.2f}"
    
    def aumentar_salario(self, percentual: float):
        """Aumenta salário - O(1)"""
        self.salario *= (1 + percentual / 100)          # O(1) - cálculo e atribuição

def exemplo_classe_simples():
    """Exemplo com classe/struct em Python"""
    print("\n=== CLASSE/STRUCT ===")
    
    # Criação - O(1)
    func1 = Funcionario("Ana Silva", 28, 3500.50)       # O(1) - criação do objeto
    print(f"Funcionário criado: {func1}")               # O(1) - chamada do __str__
    
    # Acesso aos atributos - O(1)
    print(f"Nome: {func1.nome}")                        # O(1) - acesso direto
    print(f"Idade: {func1.idade}")                      # O(1) - acesso direto
    print(f"Salário: R$ {func1.salario:.2f}")           # O(1) - acesso direto
    
    # Modificação - O(1)
    func1.aumentar_salario(10)                           # O(1) - chamada de método
    print(f"Após aumento: {func1}")

def exemplo_lista_objetos():
    """Lista de objetos (array de structs)"""
    print("\n=== LISTA DE OBJETOS ===")
    
    # Criação da lista - O(n)
    empresa = [                                          # O(n) - lista com n funcionários
        Funcionario("João", 30, 4000.0),                # O(1) - cada criação
        Funcionario("Maria", 25, 3800.0),               # O(1)
        Funcionario("Pedro", 35, 4500.0)                # O(1)
    ]
    
    # Exibir todos - O(n)
    print("Lista de funcionários:")
    for i, func in enumerate(empresa):                   # O(n) - percorre lista
        print(f"{i+1}. {func}")                         # O(1) - impressão
    
    # Encontrar maior salário - O(n)
    func_maior_salario = max(empresa, key=lambda f: f.salario)  # O(n) - percorre comparando
    print(f"Maior salário: {func_maior_salario}")

# ====================================
# 3. REFERÊNCIAS (Similar a ponteiros)
# ====================================

def exemplo_referencias_simples():
    """Demonstra como funcionam referências em Python"""
    print("\n=== REFERÊNCIAS ===")
    
    # Lista original
    lista_original = [1, 2, 3, 4, 5]                    # O(n) - criação
    print(f"Lista original: {lista_original}")
    
    # Referência (como ponteiro)
    lista_ref = lista_original                           # O(1) - cópia de referência
    print(f"Referência: {lista_ref}")
    
    # Modificação via referência - O(1)
    lista_ref[2] = 999                                   # O(1) - alteração por índice
    print(f"Após alterar via referência:")
    print(f"Lista original: {lista_original}")          # Mudou também!
    print(f"Referência: {lista_ref}")
    
    # Demonstrar que são o mesmo objeto
    print(f"São o mesmo objeto? {lista_original is lista_ref}")  # O(1) - comparação de identidade

def trocar_valores_funcao(lista: List[int], pos1: int, pos2: int):
    """Troca valores em posições específicas - O(1)"""
    print(f"Antes da troca: {lista}")
    
    # Troca usando tupla (Pythônico) - O(1)
    lista[pos1], lista[pos2] = lista[pos2], lista[pos1] # O(1) - troca atômica
    
    print(f"Após a troca: {lista}")

def exemplo_mutabilidade():
    """Demonstra mutabilidade vs imutabilidade"""
    print("\n=== MUTABILIDADE ===")
    
    # Lista mutável
    numeros = [1, 2, 3]                                  # O(1) - lista pequena
    numeros_ref = numeros                                # O(1) - referência
    
    numeros.append(4)                                    # O(1) - adiciona elemento
    print(f"Lista após append: {numeros}")
    print(f"Referência também mudou: {numeros_ref}")
    
    # String imutável
    texto = "Hello"                                      # O(1) - string pequena
    texto_ref = texto                                    # O(1) - referência
    
    texto = texto + " World"                             # O(n) - cria nova string
    print(f"String após concatenação: '{texto}'")
    print(f"Referência não mudou: '{texto_ref}'")

# ====================================
# 4. DICIONÁRIOS (Hash Tables)
# ====================================

def exemplo_dicionario_simples():
    """Dicionário como hash table"""
    print("\n=== DICIONÁRIO (HASH TABLE) ===")
    
    # Criação - O(n)
    funcionarios = {                                     # O(n) - n pares chave-valor
        "001": "Ana Silva",                              # O(1) - cada inserção
        "002": "João Santos", 
        "003": "Maria Oliveira"
    }
    
    # Acesso - O(1) médio
    codigo = "002"
    if codigo in funcionarios:                           # O(1) - verificação de chave
        nome = funcionarios[codigo]                      # O(1) - acesso por chave
        print(f"Funcionário {codigo}: {nome}")
    
    # Inserção - O(1) médio
    funcionarios["004"] = "Pedro Costa"                  # O(1) - nova entrada
    print(f"Dicionário após inserção: {funcionarios}")
    
    # Iteração - O(n)
    print("Todos os funcionários:")
    for codigo, nome in funcionarios.items():           # O(n) - percorre todos
        print(f"  {codigo}: {nome}")                    # O(1) - impressão

# ====================================
# 5. MEDIÇÃO DE PERFORMANCE
# ====================================

def medir_performance_busca():
    """Compare busca em lista vs dicionário"""
    print("\n=== COMPARAÇÃO DE PERFORMANCE ===")
    
    tamanho = 10000
    
    # Criar lista grande - O(n)
    lista_grande = list(range(tamanho))                  # O(n) - lista com n elementos
    
    # Criar dicionário grande - O(n)
    dict_grande = {i: f"valor_{i}" for i in range(tamanho)}  # O(n) - n inserções
    
    # Busca na lista - O(n)
    inicio = time.time()                                 # O(1) - marca tempo
    resultado_lista = 9999 in lista_grande              # O(n) - busca linear
    tempo_lista = time.time() - inicio                  # O(1) - calcula tempo
    
    # Busca no dicionário - O(1)
    inicio = time.time()                                 # O(1) - marca tempo
    resultado_dict = 9999 in dict_grande               # O(1) - busca hash
    tempo_dict = time.time() - inicio                   # O(1) - calcula tempo
    
    print(f"Busca em lista de {tamanho} elementos:")
    print(f"  Resultado: {resultado_lista}")
    print(f"  Tempo: {tempo_lista:.6f} segundos")
    
    print(f"Busca em dicionário de {tamanho} elementos:")
    print(f"  Resultado: {resultado_dict}")
    print(f"  Tempo: {tempo_dict:.6f} segundos")
    
    if tempo_lista > 0 and tempo_dict > 0:
        speedup = tempo_lista / tempo_dict
        print(f"  Dicionário é {speedup:.1f}x mais rápido!")

# ====================================
# FUNÇÃO PRINCIPAL
# ====================================

def main():
    """Executa todos os exemplos"""
    print("DEMONSTRAÇÃO DE ESTRUTURAS DE DADOS EM PYTHON")
    print("=" * 50)
    
    inicio_programa = time.time()                        # O(1) - marca início
    
    # Executar exemplos
    exemplo_lista_basica()                               # O(n) - listas
    operacoes_lista_simples()                            # O(n) - operações
    exemplo_classe_simples()                             # O(1) - classes
    exemplo_lista_objetos()                              # O(n) - lista de objetos
    exemplo_referencias_simples()                        # O(n) - referências
    
    # Teste de troca
    print("\n=== TESTE DE TROCA ===")
    minha_lista = [10, 20, 30, 40, 50]                  # O(1) - lista pequena
    trocar_valores_funcao(minha_lista, 1, 3)             # O(1) - troca
    
    exemplo_mutabilidade()                               # O(n) - strings
    exemplo_dicionario_simples()                         # O(n) - dicionários
    medir_performance_busca()                            # O(n) - comparação
    
    tempo_total = time.time() - inicio_programa          # O(1) - calcula tempo
    
    print("\n" + "=" * 50)
    print(f"Tempo total de execução: {tempo_total:.6f} segundos")
    print("Programa finalizado com sucesso!")

if __name__ == "__main__":
    main()

"""
RESUMO DE COMPLEXIDADES EM PYTHON:

Listas:
- Acesso por índice: O(1)
- Busca: O(n)
- Append: O(1) amortizado
- Insert no meio: O(n)
- Pop do final: O(1)
- Pop do meio: O(n)

Dicionários:
- Acesso por chave: O(1) médio
- Inserção: O(1) médio
- Remoção: O(1) médio
- Busca de chave: O(1) médio

Classes/Objetos:
- Acesso a atributo: O(1)
- Criação: O(1) + complexidade do construtor
- Método simples: O(1)

Strings (Imutáveis):
- Acesso por índice: O(1)
- Concatenação: O(n) - cria nova string
- Busca: O(n)

Principais diferenças do C:
- Python gerencia memória automaticamente
- Não há ponteiros explícitos, apenas referências
- Listas são dinâmicas por padrão
- Dicionários são hash tables built-in eficientes
"""

File: exemplos/c/estruturas_dados/estruturas_dados_simples.c
/*
 * ESTRUTURAS DE DADOS SIMPLES - Exemplos Práticos
 * Homogêneas, Heterogêneas e Ponteiros
 * 
 * Cada linha comentada com sua complexidade
 * Códigos simples e didáticos
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// ====================================
// 1. ESTRUTURAS HOMOGÊNEAS (ARRAYS)
// ====================================

/* Exemplo 1: Array simples de inteiros */
void exemplo_array_basico() {
    printf("\n=== ARRAY HOMOGÊNEO ===\n");
    
    int numeros[5];                                      // O(1) - alocação estática
    
    // Preenchimento - O(n)
    printf("Preenchendo array:\n");
    for (int i = 0; i < 5; i++) {                        // O(n) - loop executa 5 vezes
        numeros[i] = (i + 1) * 10;                       // O(1) - cálculo e atribuição
        printf("numeros[%d] = %d\n", i, numeros[i]);    // O(1) - impressão
    }
    
    // Acesso direto - O(1)
    printf("\nAcesso direto:\n");
    int valor = numeros[2];                              // O(1) - acesso por índice
    printf("Elemento na posição 2: %d\n", valor);       // O(1) - impressão
    
    // Busca linear - O(n)
    printf("\nBuscando valor 30:\n");
    int procurado = 30;
    int encontrado = -1;
    for (int i = 0; i < 5; i++) {                        // O(n) - pior caso: percorre tudo
        if (numeros[i] == procurado) {                   // O(1) - comparação
            encontrado = i;                              // O(1) - atribuição
            break;                                       // O(1) - saída do loop
        }
    }
    
    if (encontrado != -1) {                              // O(1) - verificação
        printf("Valor %d encontrado na posição %d\n", procurado, encontrado);
    } else {
        printf("Valor %d não encontrado\n", procurado);
    }
}

/* Exemplo 2: Operações com array */
int somar_array(int arr[], int tamanho) {
    printf("\nSomando elementos do array:\n");
    
    int soma = 0;                                        // O(1) - inicialização
    
    for (int i = 0; i < tamanho; i++) {                  // O(n) - loop n vezes
        soma += arr[i];                                  // O(1) - soma e acesso
        printf("Somando arr[%d] = %d, total = %d\n", i, arr[i], soma);
    }
    
    return soma;                                         // O(1) - retorno
    // Complexidade total: O(n)
}

// ====================================
// 2. ESTRUTURAS HETEROGÊNEAS (STRUCTS)
// ====================================

/* Definição de struct simples */
typedef struct {
    char nome[30];        // Array de caracteres
    int idade;           // Inteiro
    float salario;       // Número decimal
} Funcionario;

/* Exemplo 3: Criando e usando struct */
void exemplo_struct_basico() {
    printf("\n=== STRUCT HETEROGÊNEA ===\n");
    
    // Criação - O(1) para campos numéricos, O(n) para strings
    Funcionario func1;                                   // O(1) - aloca espaço na stack
    
    strcpy(func1.nome, "Ana Silva");                     // O(n) - copia string (n = tamanho)
    func1.idade = 28;                                    // O(1) - atribuição direta
    func1.salario = 3500.50;                             // O(1) - atribuição direta
    
    // Exibição - O(1) para acessos diretos
    printf("Dados do funcionário:\n");
    printf("Nome: %s\n", func1.nome);                   // O(1) - acesso direto ao campo
    printf("Idade: %d anos\n", func1.idade);            // O(1) - acesso direto
    printf("Salário: R$ %.2f\n", func1.salario);        // O(1) - acesso direto
    
    // Modificação - O(1)
    func1.salario = func1.salario * 1.10;               // O(1) - cálculo e atribuição
    printf("Novo salário (aumento 10%%): R$ %.2f\n", func1.salario);
}

/* Exemplo 4: Array de structs */
void exemplo_array_structs() {
    printf("\n=== ARRAY DE STRUCTS ===\n");
    
    Funcionario empresa[3];                              // O(1) - aloca array de structs
    
    // Preenchimento - O(n×m) onde n=funcionários, m=tamanho médio do nome
    strcpy(empresa[0].nome, "João");                     // O(m) - copia string
    empresa[0].idade = 30;                               // O(1)
    empresa[0].salario = 4000.0;                         // O(1)
    
    strcpy(empresa[1].nome, "Maria");                    // O(m)
    empresa[1].idade = 25;                               // O(1)
    empresa[1].salario = 3800.0;                         // O(1)
    
    strcpy(empresa[2].nome, "Pedro");                    // O(m)
    empresa[2].idade = 35;                               // O(1)
    empresa[2].salario = 4500.0;                         // O(1)
    
    // Exibir todos - O(n)
    printf("Lista de funcionários:\n");
    for (int i = 0; i < 3; i++) {                        // O(n) - loop pelos funcionários
        printf("%d. %s, %d anos, R$ %.2f\n", 
               i+1, empresa[i].nome, empresa[i].idade, empresa[i].salario); // O(1) - acesso e impressão
    }
    
    // Buscar funcionário mais bem pago - O(n)
    int indice_maior_salario = 0;                        // O(1) - inicialização
    for (int i = 1; i < 3; i++) {                        // O(n) - percorre funcionários
        if (empresa[i].salario > empresa[indice_maior_salario].salario) { // O(1) - comparação
            indice_maior_salario = i;                    // O(1) - atualização
        }
    }
    
    printf("Maior salário: %s (R$ %.2f)\n", 
           empresa[indice_maior_salario].nome, empresa[indice_maior_salario].salario);
}

// ====================================
// 3. PONTEIROS
// ====================================

/* Exemplo 5: Ponteiros básicos */
void exemplo_ponteiros_simples() {
    printf("\n=== PONTEIROS BÁSICOS ===\n");
    
    int numero = 42;                                     // O(1) - cria variável
    int* ptr = &numero;                                  // O(1) - ptr guarda endereço de numero
    
    printf("Valor original: %d\n", numero);             // O(1) - acesso direto
    printf("Endereço da variável: %p\n", &numero);      // O(1) - pega endereço
    printf("Valor do ponteiro: %p\n", ptr);             // O(1) - mostra endereço armazenado
    printf("Valor via ponteiro: %d\n", *ptr);           // O(1) - desreferenciamento
    
    // Alteração via ponteiro - O(1)
    *ptr = 100;                                          // O(1) - altera valor via ponteiro
    printf("Novo valor: %d\n", numero);                 // O(1) - numero foi alterado!
}

/* Exemplo 6: Função que troca valores usando ponteiros */
void trocar_valores(int* a, int* b) {
    printf("Trocando valores via ponteiros:\n");
    printf("Antes: *a = %d, *b = %d\n", *a, *b);       // O(1) - acesso via ponteiro
    
    int temp = *a;                                       // O(1) - copia valor de a
    *a = *b;                                             // O(1) - a recebe valor de b
    *b = temp;                                           // O(1) - b recebe valor original de a
    
    printf("Depois: *a = %d, *b = %d\n", *a, *b);      // O(1) - verificação
    // Complexidade total: O(1)
}

/* Exemplo 7: Alocação dinâmica simples */
void exemplo_alocacao_dinamica() {
    printf("\n=== ALOCAÇÃO DINÂMICA ===\n");
    
    int tamanho = 5;
    printf("Criando array dinâmico de %d elementos:\n", tamanho);
    
    // Alocação - O(1)
    int* arr = malloc(tamanho * sizeof(int));            // O(1) - aloca memória
    
    if (arr == NULL) {                                   // O(1) - verificação de erro
        printf("Erro: Não foi possível alocar memória!\n");
        return;
    }
    
    // Preenchimento - O(n)
    for (int i = 0; i < tamanho; i++) {                  // O(n) - loop n vezes
        arr[i] = i * i;  // Quadrado do índice            // O(1) - cálculo e atribuição
        printf("arr[%d] = %d\n", i, arr[i]);            // O(1) - impressão
    }
    
    // Soma usando aritmética de ponteiros - O(n)
    int soma = 0;                                        // O(1) - inicialização
    for (int i = 0; i < tamanho; i++) {                  // O(n) - loop
        soma += *(arr + i);  // Equivale a arr[i]        // O(1) - aritmética + acesso
        printf("Somando arr[%d] = %d, total = %d\n", i, *(arr + i), soma);
    }
    
    printf("Soma total: %d\n", soma);                   // O(1) - resultado
    
    // IMPORTANTE: Liberar memória - O(1)
    free(arr);                                           // O(1) - libera memória
    arr = NULL;                                          // O(1) - boa prática
    printf("Memória liberada com sucesso!\n");
}

/* Exemplo 8: Ponteiro para struct */
void exemplo_ponteiro_struct() {
    printf("\n=== PONTEIRO PARA STRUCT ===\n");
    
    // Criar funcionário - O(1) + O(n) para string
    Funcionario* emp = malloc(sizeof(Funcionario));      // O(1) - aloca struct dinamicamente
    
    if (emp == NULL) {                                   // O(1) - verificação
        printf("Erro de alocação!\n");
        return;
    }
    
    // Preenchimento usando ponteiro - O(1) + O(n) para string
    strcpy(emp->nome, "Carlos Santos");                  // O(n) - copia string
    emp->idade = 40;                                     // O(1) - acesso via seta
    emp->salario = 5000.0;                               // O(1) - acesso via seta
    
    // Exibição - O(1) para cada campo
    printf("Dados via ponteiro:\n");
    printf("Nome: %s\n", emp->nome);                     // O(1) - acesso direto
    printf("Idade: %d\n", emp->idade);                   // O(1) - acesso direto
    printf("Salário: R$ %.2f\n", emp->salario);          // O(1) - acesso direto
    
    // Modificação via ponteiro - O(1)
    emp->salario += 500.0;  // Aumento                   // O(1) - operação e atribuição
    printf("Novo salário: R$ %.2f\n", emp->salario);     // O(1) - verificação
    
    // Liberação - O(1)
    free(emp);                                           // O(1) - libera struct
    emp = NULL;                                          // O(1) - anula ponteiro
}

// ====================================
// FUNÇÃO PRINCIPAL
// ====================================

int main() {
    printf("DEMONSTRAÇÃO DE ESTRUTURAS DE DADOS\n");
    printf("==========================================\n");
    
    // Medir tempo total de execução
    clock_t inicio_programa = clock();                   // O(1) - marca tempo inicial
    
    // Executar exemplos
    exemplo_array_basico();                              // O(n) - arrays
    exemplo_struct_basico();                             // O(n) - structs  
    exemplo_array_structs();                             // O(n×m) - array de structs
    exemplo_ponteiros_simples();                         // O(1) - ponteiros básicos
    
    // Exemplo com troca de valores
    printf("\n=== TESTE DE TROCA ===\n");
    int x = 10, y = 20;                                  // O(1) - inicialização
    printf("Valores originais: x = %d, y = %d\n", x, y);
    trocar_valores(&x, &y);                              // O(1) - troca via ponteiros
    printf("Valores após troca: x = %d, y = %d\n", x, y);
    
    exemplo_alocacao_dinamica();                         // O(n) - alocação dinâmica
    exemplo_ponteiro_struct();                           // O(n) - ponteiro para struct
    
    // Tempo total
    clock_t fim_programa = clock();                      // O(1) - marca tempo final
    double tempo_total = ((double)(fim_programa - inicio_programa)) / CLOCKS_PER_SEC;
    
    printf("\n==========================================\n");
    printf("Tempo total de execução: %.6f segundos\n", tempo_total);
    printf("Programa finalizado com sucesso!\n");
    
    return 0;                                            // O(1) - retorno
}

/*
 * RESUMO DE COMPLEXIDADES:
 * 
 * Arrays:
 * - Acesso: O(1)
 * - Busca: O(n)
 * - Preenchimento: O(n)
 * 
 * Structs:
 * - Acesso a campo: O(1)
 * - Cópia de string: O(n)
 * - Comparação: O(1) para números, O(n) para strings
 * 
 * Ponteiros:
 * - Declaração: O(1)
 * - Atribuição: O(1)
 * - Desreferenciamento: O(1)
 * - Aritmética: O(1)
 * 
 * Alocação Dinâmica:
 * - malloc(): O(1)
 * - free(): O(1)
 * - Preenchimento: O(n)
 */
